[TorchElastic] Add option to configure log prefix for each rank

kurman opened the pull request:
Summary:
Add an ability to customize log lines and addtional template like behavior to enrich log information.

Motivation:
a) Log stream processing/aggregation gains additional value when it includes information about the global rank. Extension to that is that it will be easier to map ranks to hosts from log stream information (less relevant at the moment)
b) Users can easily map the failure to the right rank without matching node rank offset+local rank.

Implementation
- BC change - keeps the logs line prefix as `[<role name><local rank>]:`
- Optional env variable TORCHELASTIC_LOG_LINE_HEADER that will be used as a prefix when specified and currently exposes `role_name`, `rank` and `local_rank` variables that will be bound when agent assigns the ranks.

Test Plan:
CI

https://fburl.com/mlhub/mzx5xspv

Differential Revision: D50584590

kurman commented:
@kiukchung @d4l3k do you mind helping merging the changes?

d4l3k commented on torch/distributed/elastic/multiprocessing/tail_log.py:
```diff
@@ -111,10 +113,13 @@ def start(self) -> "TailLog":
 
         for local_rank, file in self._log_files.items():
+            header = f"[{self._name}{local_rank}]:"
+            if self._log_line_headers and local_rank in self._log_line_headers:
```
Is the intention to use the default if an empty dict is passed? can we document that behavior?

d4l3k commented on torch/distributed/elastic/multiprocessing/api.py:
```diff
@@ -622,6 +626,7 @@ def __init__(
             tee_stdouts,
             tee_stderrs,
             error_files,
+            log_line_headers
```
nit: use kwargs for this many arguments

d4l3k reviewed (commented):
I don't have any big objections but is there a reason we don't just make the more expressive the default? If we want backwards compatibility we can have an env to switch to previous behavior. I can't think of many use cases for having full templating support

kurman commented:
> more expressive the default?

printing global rank? And I think we need to be maintain existing setup so we need to opt-in into a new behavior.

kiukchung commented on torch/distributed/elastic/agent/server/local_elastic_agent.py:
```diff
@@ -254,6 +263,17 @@ def _start_workers(self, worker_group: WorkerGroup) -> Dict[int, Any]:
             if "OMP_NUM_THREADS" in os.environ:
                 worker_env["OMP_NUM_THREADS"] = os.environ["OMP_NUM_THREADS"]
 
+            if "TORCHELASTIC_LOG_LINE_HEADER" in os.environ:
```
for the sake of keeping the elastic agent programmatic, this be taken as a ctor argument. The `TORCHELASTIC_LOG_LINE_HEADER` should be parsed in the `torchrun` CLI (e.g. `torch/distributed/run.py`)

kurman replied:
Good suggestion, done.

kiukchung commented on torch/distributed/elastic/multiprocessing/__init__.py:
```diff
@@ -88,6 +88,7 @@ def start_processes(
     args: Dict[int, Tuple],
     envs: Dict[int, Dict[str, str]],
     log_dir: str,
+    log_line_headers: Dict[int, str],
```
to make programmatic usages BC, `log_line_headers` needs to be made `Optional[Dict[int, str]]` and defaulted to `None`.

kiukchung commented on torch/distributed/elastic/multiprocessing/tail_log.py:
```diff
@@ -86,6 +86,7 @@ def __init__(
         name: str,
         log_files: Dict[int, str],
         dst: TextIO,
+        log_line_headers: Dict[int, str],
```
1. Same here, for API BC, make `log_line_headers: Optional[Dict[int, str]] = None`.
2. Since all local procs should have the same prefix template, it might be cleaner to pass a map of `local_rank -> rank` and a string template `log_line_prefix` and do the substitution locally in `TailLog`, rather than making the API take a per-local-rank header.

kiukchung commented on torch/distributed/elastic/multiprocessing/tail_log.py:
```diff
@@ -62,7 +62,7 @@ class TailLog:
 
      log_files = {0: "/tmp/0_stdout.log", 1: "/tmp/1_stdout.log"}
-     tailer = TailLog("trainer", log_files, sys.stdout).start()
+     tailer = TailLog("trainer", log_files, sys.stdout, {}).start()
```
you can leave this documentation as is by making the ctor of `TailLog` BC by defaulting `log_line_headers` to `None` , then per @d4l3k 's point below, add an extra paragraph to this docstring on how to customize the log line prefix.

kiukchung commented on torch/distributed/elastic/multiprocessing/api.py:
```diff
@@ -388,6 +389,7 @@ def __init__(
         tee_stdouts: Dict[int, str],
         tee_stderrs: Dict[int, str],
         error_files: Dict[int, str],
+        log_line_headers: Dict[int, str],
```
nit: can we make these `log_line_prefix`? I know I used `header` in `tail_logfile()` but in various documentations we refer to this feature as "localrank log prefixing"

kiukchung commented on torch/distributed/elastic/agent/server/local_elastic_agent.py:
```diff
@@ -254,6 +263,17 @@ def _start_workers(self, worker_group: WorkerGroup) -> Dict[int, Any]:
+            if "TORCHELASTIC_LOG_LINE_HEADER" in os.environ:
+                log_line_header = os.environ["TORCHELASTIC_LOG_LINE_HEADER"]
+
+                log_line_header = Template(log_line_header).safe_substitute(
+                    role_name=spec.role,
+                    rank=worker.global_rank,
+                    local_rank=local_rank,)
+                log_line_headers[local_rank] = log_line_header
```
I think this logic should move into `TailLog` because otherwise, we have two places where we default to `[$ROLE_NAME$LOCAL_RANK]:` (in the `else` branch below and in `TailLog` L116.

kiukchung commented on torch/distributed/elastic/agent/server/local_elastic_agent.py:
```diff
@@ -254,6 +263,17 @@ def _start_workers(self, worker_group: WorkerGroup) -> Dict[int, Any]:
+                    local_rank=local_rank,)
+                log_line_headers[local_rank] = log_line_header
+            else:
+                log_line_headers[local_rank] = f"[{spec.role}:{local_rank}]"
```
shouldn't this be `f"[{spec.role}{local_rank}]:"` to be BC? (see my comment about moving the substitution into `TailLog`)

kiukchung commented:
Thanks for doing this, its been a longtime TODO for me. Overall I think rather than giving users an option to customize the prefix, we should just default it to `[{role}{global_rank}][{local_rank]}]:` (example: `[trainer16][0]:`).

Yes it breaks BC but it'll only meaningfully break BC if someone is relying on the prefix to parse logs (unless someone in Meta is doing this).

kurman commented:
> someone is relying on the prefix to parse logs

Logs are being parsed now so once we migrate to new format, we can then make your recommended pattern a default.

kiukchung commented on torch/distributed/elastic/agent/server/local_elastic_agent.py:
````diff
@@ -79,6 +80,13 @@ class LocalElasticAgent(SimpleElasticAgent):
     named pipe that ```LocalElasticAgent``` uses.
 
+    Logs are written to the specified log directory. Each log line will be by default
+    prefixed by "[<ROLE NAME><LOCAL_RANK>]:" string unless `log_line_prefix_template`
````
nit:  a better way to document template pattern is by making this:

```
prefixed by ``[${role_name}${local_rank}]:`` (e.g. ``[trainer0]: foobar``). 
Log prefixes can be customized by passing a `template string <https://docs.python.org/3/library/string.html#template-strings>`_ as the ``log_line_prefix_template`` argument. 
The following macros (identifiers) are substituted at runtime: ``${role_name}, ${local_rank}, ${rank}``.
 For example, to prefix each log line with global rank instead of the local rank, set ``log_line_prefix_template = "[${rank}]:".
```

kiukchung commented on test/distributed/elastic/agent/server/test/local_elastic_agent_test.py:
```diff
@@ -333,6 +338,7 @@ def run_agent(
         master_port_override: Optional[int] = None,
         is_host=True,
         monitor_interval=0.01,
+        log_line_prefix_template: Optional[str] = None,
```
nit: we should be consistent between the variable name and the env var. Since the env var was called `TORCHELASTIC_LOG_PREFIX_TEMPLATE` lets call this variable `log_prefix_template`

kurman replied:
Changed the env var name to TORCHELASTIC_LOG_LINE_PREFIX_TEMPLATE. It is more expressive, esp at `TailLog` scope.

kiukchung commented on torch/distributed/run.py:
```diff
@@ -713,6 +713,10 @@ def config_from_args(args) -> Tuple[LaunchConfig, Union[Callable, str], List[str
         # This env variable will be passed down to the subprocesses
         os.environ["OMP_NUM_THREADS"] = str(omp_num_threads)
 
+    log_line_prefix_template = "[${role}:${local_rank}]"
+    if "TORCHELASTIC_LOG_PREFIX_TEMPLATE" in os.environ:
+        log_line_prefix_template = os.environ["TORCHELASTIC_LOG_PREFIX_TEMPLATE"]
```
this can be written more succinctly as:

```
log_prefix_template = os.getenv("TORCHELASTIC_LOG_PREFIX_TEMPLATE")
```

let `log_prefix_template` be `None` and get defaulted to `[${role_name}:${local_rank}]:` in `TailLogs`

kurman replied:
For the default template - keeping it local_elastic_agent (See [reasoning](https://github.com/pytorch/pytorch/pull/112357#discussion_r1381156896))

kiukchung commented on torch/distributed/run.py:
```diff
@@ -713,6 +713,10 @@ def config_from_args(args) -> Tuple[LaunchConfig, Union[Callable, str], List[str
         # This env variable will be passed down to the subprocesses
         os.environ["OMP_NUM_THREADS"] = str(omp_num_threads)
 
+    log_line_prefix_template = "[${role}:${local_rank}]"
```
I'm confused at what the default prefix format should be. Is it
`[trainer1]:` or `[trainer:1]`?

If the former, shouldn't this be `[${role}${local_rank}]:`? Otherwise this would prefix `[trainer:1]<log msg>` instead of `[trainer1]:<log msg>`

kiukchung commented on test/distributed/elastic/multiprocessing/tail_log_test.py:
```diff
@@ -53,7 +53,7 @@ def test_tail(self):
 
         dst = io.StringIO()
-        tail = TailLog("writer", log_files, dst, interval_sec).start()
+        tail = TailLog("writer", log_files, dst, interval_sec=interval_sec).start()
```
is this change intentional? (using kwargs for `interval_sec` only)

kiukchung commented on torch/distributed/elastic/agent/server/local_elastic_agent.py:
```diff
@@ -254,6 +265,17 @@ def _start_workers(self, worker_group: WorkerGroup) -> Dict[int, Any]:
                 worker_env["OMP_NUM_THREADS"] = os.environ["OMP_NUM_THREADS"]
 
+
+            prefix_template = "[${role}${local_rank}]:"
```
1. we should really have one place where we default to this prefix template. currently the prefix defaults in 3 places: 
     1. torch.distributed.run (to a sightly different format)
     2. here
     3. TailLogs (we should default it here and you can do this by passing the local_rank -> global rank mapping to TailLogs instead of the substituted prefix for each local rank)
    

5. nit: lets be consistent with the variable name, so this should be `log_line_prefix_template`).

kurman replied:
I decided to keep it here for two reasons:
1. We can limit the information passed AND most of the contextual data is here, so adding new variables will be simpler
2. I would like to make sure that we always pass header prefix mapping once this lands and we don't run into issues.

kiukchung reviewed (commented):
Thanks for making some changes. My main concerns are around:

1. Single place for prefix default
2. Being more consistent with naming (log_prefix vs log_line_prefix_template, vs log_prefix_template)
3. Seemingly different prefix defaults between torch.distributed.run and (elastic_agent,  TailLogs)

kurman committed:
[TorchElastic] Add option to configure log prefix for each rank (#112357)

Summary:

Add an ability to customize log lines and addtional template like behavior to enrich log information.

Motivation:
a) Log stream processing/aggregation gains additional value when it includes information about the global rank. Extension to that is that it will be easier to map ranks to hosts from log stream information (less relevant at the moment)
b) Users can easily map the failure to the right rank without matching node rank offset+local rank.

Implementation
- BC change - keeps the logs line prefix as "[<role name><local rank>]:"
- Optional env variable TORCHELASTIC_LOG_LINE_HEADER that will be used as a prefix when specified and currently exposes `role_name`, `rank` and `local_rank` variables that will be bound when agent assigns the ranks.

Test Plan:
CI

https://fburl.com/mlhub/mzx5xspv

Differential Revision: D50584590

kurman commented:
> * Single place for prefix default
> * Being more consistent with naming (log_prefix vs log_line_prefix_template, vs log_prefix_template)
> * Seemingly different prefix defaults between torch.distributed.run and (elastic_agent,  TailLogs)

Addressed all of them. One thing that I believe is outstanding is making log_line_prefix_headers non-optional, but that we can follow up in different less risky PR as a refactoring change.

kurman commented:
@pytorchbot merge
